### PR TITLE
feat: ACI-2805 Add flag to follow xcode symlinks

### PR DESCRIPTION
**Pull request by zsolt-marta-bitrise**

### CODE REVIEW

File: cmd/saveXcodeDerivedDataFiles.go [hunk 1]
```diff
@@ -49,7 +50,18 @@ var saveXcodeDerivedDataFilesCmd = &cobra.Command{
...
-		op, cmdError := saveXcodeDerivedDataFilesCmdFn(cmd.Context(), authConfig, CacheMetadataPath, projectRoot, cacheKey, ddPath, xcodeCachePath, logger, tracker, startT, os.Getenv)
+		op, cmdError := saveXcodeDerivedDataFilesCmdFn(cmd.Context(),
+			authConfig,
+			CacheMetadataPath,
+			projectRoot,
+			cacheKey,
+			ddPath,
+			xcodeCachePath,
+			followSymlinks,
+			logger,
```
Comment (zsolt-marta-bitrise) [inline cmd/saveXcodeDerivedDataFiles.go:61]: Reminder https://bitrise.atlassian.net/browse/ACI-2397

File: internal/xcode/file_group_info.go [hunk 2]
```diff
@@ -103,29 +110,60 @@ func collectFileGroupInfo(cacheDirPath string, rootDir string, collectAttributes
...
 
 		return nil
 	}
 
-	if inf.Mode()&os.ModeSymlink != 0 {
-		logger.Debugf("Skipping symbolic link: %s", path)
-
-		return nil
+	if fileInfo.Mode()&os.ModeSymlink != 0 {
+		return collectSymlink(path, fgi, followSymlinks, rootDir, logger)
```
Comment (imrekel) [inline internal/xcode/file_group_info.go:166]: I don't get it, if a file is symlink, we call `collectSymlink` on it. But `collectSymlink` will call `collectFileMetadata` on the file if it's not a directory. Do I miss something on how the circular method calls are broken?

Comment (zsolt-marta-bitrise) [reply internal/xcode/file_group_info.go:166]: `collectSymlink()` will call `resolveSymlink()` to resolve the symlink and collect the file info on that target. The resolution is called recursively until we find a regular file, so `collectFileMetadata()` is called again with the target files only

File: internal/xcode/file_group_info.go [hunk 3]
```diff
@@ -197,6 +286,42 @@ func setAttributes(path string, attributes map[string]string) error {
 	return nil
 }
 
+func restoreSymlink(symlink SymlinkInfo, forceOverwrite bool, logger log.Logger) bool {
+	fileInfo, err := os.Lstat(symlink.Path)
+	if err == nil && fileInfo.Mode()&os.ModeSymlink != 0 {
+		if !forceOverwrite {
```
Comment (imrekel) [inline internal/xcode/file_group_info.go:292]: just a note: currently we overwrite regular files by default (even if forceOverwrite=false) - forceOverwrite=true means that we even try to overWrite files with read only file mode

so I'd remove this check and overwrite symlinks by default too - we will need a new config option to skip existing files and if that's set then I'd skip symlinks too

File: cmd/saveXcodeDerivedDataFiles.go [hunk 4]
```diff
@@ -86,10 +98,21 @@ func init() {
 		panic(err)
 	}
 	saveXcodeDerivedDataFilesCmd.Flags().String("xcodecache-path", "", "Path to the Xcode cache directory folder to be saved. If not set, it will not be uploaded.")
+	saveXcodeDerivedDataFilesCmd.Flags().Bool("follow-symlinks", false, "Follow symlinks when calculating metadata (default: false)")
```
Comment (imrekel) [inline cmd/saveXcodeDerivedDataFiles.go:101]:
```suggestion
	saveXcodeDerivedDataFilesCmd.Flags().Bool("follow-symlinks", false, "Follow symlinks when calculating metadata and saving files to the cache (default: false)")
```
```suggestion
	saveXcodeDerivedDataFilesCmd.Flags().Bool("follow-symlinks", false, "Follow symlinks when calculating metadata (default: false)")
```

**Review by imrekel — APPROVED** [review]

**Pull request merged**